File: my-folder/0437-path-sum-iii/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public int pathSum(TreeNode root, int targetSum) {
        if(root==null){
            return 0;
        }
       long temp=targetSum;
 

       return  helper(root,temp)+ pathSum(root.left,targetSum)+ pathSum(root.right,targetSum);
    } 
    public int helper(TreeNode root, long target){
       
        if(root==null){
            return 0;
        } 
       
        int ans=0;
        if(target==(long)root.val){
            ans++;
        }
         ans+=helper(root.left,target-(long)root.val); 
         ans+=helper(root.right,target-(long)root.val);
       
        return ans;
    }
}

File: my-folder/0437-path-sum-iii/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        def dfs(node, target, curr, dic):
            if not node:
                return 

            curr += node.val
            self.res += dic[curr-target]
            dic[curr] += 1

            dfs(node.left, target, curr, dic)
            dfs(node.right, target, curr, dic)

            dic[curr] -= 1
        self.res = 0
        dic = defaultdict(int)
        dic[0] += 1
        dfs(root, targetSum, 0, dic)
        return self.res
